### PR TITLE
luci-app-babeld: rewrite javascript of @PolynomialDivision version

### PR DESCRIPTION
**Pull request by JonnyTischbein**
In addition to https://github.com/openwrt/luci/pull/4780 I rewrote the javascript code.

Updates are:

- no external CDN (was only used for testing)
- luci css (bootstrap) instead of external bootstrap
- use the div classed table instead of table
- no jquery, written with javascript DOM
- javascript functions outsource to resource file, only ubus call and render function calls remain in htm
- only one render function per category

Thanks again to @PolynomialDivision for the help!

image-123-1

image-123

[Added IPC Javascript Basics]
Signed-off-by: Nick Hainke <vincent@systemli.org>
[Rewrote Javascript]
Signed-off-by: Jonathan Pagel <jonny_tischbein@systemli.org>

### CODE REVIEW

**Comment by PolynomialDivision** [comment]
@aparcar ;)

File: applications/luci-app-babeld/root/www/luci-static/resources/babeld.js [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    neighbourRow.setAttribute('class', 'tr');
+    var neighbourContent = '<div class="td" data-title="info_babeld-version">' + data["babeld-version"] + '</div>\
+                            <div class="td" data-title="info_dev">' + data["my-id"] + '</div>\
+                            <div class="td" data-title="info_hello-reach">' + data["host"] + '</div>';
+
+
+    neighbourRow.innerHTML = neighbourContent;
+    table.appendChild(neighbourRow);
+    target.appendChild(table);
+}
```
Comment (PolynomialDivision) [inline applications/luci-app-babeld/root/www/luci-static/resources/babeld.js:190]: newline missing

File: applications/luci-app-babeld/luasrc/view/babeld.htm [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    var neighbours = ubus_call("babeld", "get_neighbours", {});
+
+    renderTableInfo(info, 'babeld');
+    renderTableXRoutes(xroutes, 'babeld');
+    renderTableRoutes(routes, 'babeld');
+    renderTableNeighbours(neighbours, 'babeld');
+</script>
+
+
+
```
Comment (PolynomialDivision) [inline applications/luci-app-babeld/luasrc/view/babeld.htm:96]: to much newline :P

File: applications/luci-app-babeld/luasrc/view/babeld.htm [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+<script type="text/javascript">
+    data = {};
+	data["ubus_rpc_session"] = "<%=luci.dispatcher.context.authsession%>"
+	origin = document.location.href.replace(location.pathname, "")
+	ubus_url = origin + "/ubus/"
+
+	var info = ubus_call("babeld", "get_info", {});
+	var xroutes = ubus_call("babeld", "get_xroutes", {});
+	var routes = ubus_call("babeld", "get_routes", {});
```
Comment (PolynomialDivision) [inline applications/luci-app-babeld/luasrc/view/babeld.htm:86]: ident seems wrong?

File: applications/luci-app-babeld/root/www/luci-static/resources/babeld.js [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+            var neighbourContent = '<div class="td" data-title="' + protocol + '_neighbour">' + neighbour + '</div>\
+                                    <div class="td" data-title="neighbours_' + protocol + '_dev">' + data[protocol][neighbour]["dev"] + '</div>\
+                                    <div class="td" data-title="neighbours_' + protocol + '_hello-reach">' + data[protocol][neighbour]["hello-reach"] + '</div>\
+                                    <div class="td" data-title="neighbours_' + protocol + '_rxcost">' + data[protocol][neighbour]["rxcost"] + '</div>\
+                                    <div class="td" data-title="neighbours_' + protocol + '_txcost">' + data[protocol][neighbour]["txcost"] + '</div>\
+                                    <div class="td" data-title="neighbours_' + protocol + '_rtt">' + data[protocol][neighbour]["rtt"] + '</div>\
+                                    <div class="td" data-title="neighbours_' + protocol + '_channel">' + data[protocol][neighbour]["channel"] + '</div>\
+                                    <div class="td" data-title="neighbours_' + protocol + '_if_up">' + data[protocol][neighbour]["if_up"] + '</div>';
+
+
```
Comment (PolynomialDivision) [inline applications/luci-app-babeld/root/www/luci-static/resources/babeld.js:152]: remove empty line?

**Comment by PolynomialDivision** [comment]
@JonnyTischbein You are missing changes in `luci-app-babeld.json`.

The file has to change to:
```

{
	"luci-app-babeld": {
		"description": "Grant UCI access for babeld",
		"read": {
			"uci": [ "babeld" ],
			"ubus": {
				"babeld": [ "*" ]
			}
		},
		"write": {
			"uci": [ "babeld" ],
			"ubus": {
				"babeld": [ "*" ]
		}
		}
	}
}
```

**Comment by PolynomialDivision** [comment]
And u are missing changes in `applications/luci-app-babeld/root/usr/share/luci/menu.d/luci-app-babeld.json`

It has to be

```
{
	"admin/status/babeld": {
		"title": "Babeld",
		"action": {
			"type": "template",
			"path": "babeld"
		},
		"depends": {
			"acl": [ "luci-app-babeld" ]
		}
	}
}
```

**Comment by JonnyTischbein** [comment]
I changed them and forgot to commit :upside_down_face: 

Are the dependencies right ? 
https://github.com/JonnyTischbein/luci/blob/babeld-javascript/applications/luci-app-babeld/Makefile#L3-L4

**Comment by PolynomialDivision** [comment]
> I changed them and forgot to commit upside_down_face
> 
> Are the dependencies right ?
> https://github.com/JonnyTischbein/luci/blob/babeld-javascript/applications/luci-app-babeld/Makefile#L3-L4

Good question. We have again lua code that is checking the acl. I thought we do the javascript rewrite, since we want to drop lua at the end.

**Comment by PolynomialDivision** [comment]
@JonnyTischbein There is still something wrong. I tested it again on mips and it does not show up. I will have a closer look, soon.

**Comment by PolynomialDivision** [comment]
If I access `https://10.0.0.3/cgi-bin/luci/admin/status/babeld` it tells me `403 Forbidden` in console output.

File: applications/luci-app-babeld/root/usr/share/rpcd/acl.d/luci-app-babeld.json [hunk 5]
```diff
@@ -2,10 +2,14 @@
...
+			"ubus": {
+				"babeld": [ "*" ]
+			}
 		},
 		"write": {
-			"uci": [ "babeld" ]
+			"uci": [ "babeld" ],
+			"ubus": {
+				"babeld": [ "*" ]
 		}
```
Comment (PolynomialDivision) [inline applications/luci-app-babeld/root/usr/share/rpcd/acl.d/luci-app-babeld.json:14]: I think you are missing here a `}`.

File: applications/luci-app-babeld/root/usr/share/rpcd/acl.d/luci-app-babeld.json [hunk 6]
```diff
@@ -2,10 +2,14 @@
...
+			"uci": [ "babeld" ],
+			"ubus": {
+				"babeld": [ "*" ]
+			}
 		},
 		"write": {
-			"uci": [ "babeld" ]
+			"uci": [ "babeld" ],
+			"ubus": {
+				"babeld": [ "*" ]
```
Comment (PolynomialDivision) [inline applications/luci-app-babeld/root/usr/share/rpcd/acl.d/luci-app-babeld.json:13]: an here u are missing again `}`. ;)

**Comment by PolynomialDivision** [comment]
@JonnyTischbein The access control list is wrong. It has to be

```
{
	"luci-app-babeld": {
		"description": "Grant UCI access for babeld",
		"read": {
			"uci": [ "babeld" ],
			"ubus": {
				"babeld": [ "*" ]
			}
		},
		"write": {
			"uci": [ "babeld" ],
			"ubus": {
				"babeld": [ "*" ]
			}
		}
	}
}
```
The change fixes everything and the app works then smooth. :)

**Comment by JonnyTischbein** [comment]
I'm sorry!

This time I compilied the luci-application-babeld package and tested it in my openwrt VM without problems. Can you confirm ?

**Comment by PolynomialDivision** [comment]
Compile- and run-tested on WDR4300. Can someone merge, I don't have commit access? @aparcar @hnyman ?

**Comment by PolynomialDivision** [comment]
Maybe we should add something like

```
define Package/luci-app-babeld/postinst
#!/bin/sh
[ -n "$$IPKG_INSTROOT" ] || uci set babeld.@general[0].ubus_bindings='true'
endef
```

to the Makefile.

**Comment by JonnyTischbein** [comment]
I did fix the title linebreaks and removed the luca function ` function apply_acls(filename, session)`, because it wasn't needed in my test. I hope that is correct.

**Comment by PolynomialDivision** [comment]
> I did fix the title linebreaks and removed the luca function ` function apply_acls(filename, session)`, because it wasn't needed in my test. I hope that is correct.

Did u try a reboot and test that again? ^^

**Comment by PolynomialDivision** [comment]
Run and compiletested again on wdr4300

**Comment by PolynomialDivision** [comment]
@JonnyTischbein I think u can also remove luci-compat from dependency list in Makefile. Since we did all the rewrite to drop the lua dependencies. xD

**Comment by JonnyTischbein** [comment]
`root/etc/uci-defaults/40_luci-babeld` with `ubus_binding=true` added and tested on openwrt VM SNAPSHOT r15721-cd2b661453

**Comment by PolynomialDivision** [comment]
> `root/etc/uci-defaults/40_luci-babeld` with `ubus_binding=true` added and tested on openwrt VM `SNAPSHOT r15721-cd2b661453 `

Nice! :) @aparcar please merge. I also compile and runtested again on WDR4300.

**Pull request merged**